### PR TITLE
Created the Application Loade

### PR DESCRIPTION
**Pull request by omainegra**
Created the Application Loader, with fallback to default Guice Application Loader.

The following actions should be to remove `guice` from project and write the `Component`

### CODE REVIEW

File: web/app/TheApplicationLoader.scala [hunk 1]
```diff
@@ -0,0 +1,5 @@
+import play.api.inject.guice.GuiceApplicationLoader
+
+class TheApplicationLoader extends GuiceApplicationLoader {
```
Comment (ScalaWilliam) [inline web/app/TheApplicationLoader.scala]: On my machine, I tried:
```scala
import play.api.{Application, ApplicationLoader}
import play.api.inject.guice.{GuiceApplicationBuilder, GuiceApplicationLoader}

class TheApplicationLoader extends GuiceApplicationLoader {
  override protected def builder(context: ApplicationLoader.Context): GuiceApplicationBuilder = {
    ???
  }
}
```

And `web/it:test` passes without exception. I would expect it to fail, and indicate that the app is going through this class.

Is there any way to positively show this is being loaded?

That said, I see nothing wrong with your code, so I'm a bit puzzled.

File: project/Dependencies.scala [hunk 2]
```diff
@@ -4,10 +4,10 @@ object Dependencies {
 
   val scalatest: ModuleID = "org.scalatest" %% "scalatest" % "3.0.3"
   val scalacheck: ModuleID = "org.scalacheck" %% "scalacheck" % "1.13.5"
-  val sourcecode: ModuleID = "com.lihaoyi" %% "sourcecode" % "0.1.3"
+  val sourcecode: ModuleID = "com.lihaoyi" %% "sourcecode" % "0.1.4"
```
Comment (ScalaWilliam) [inline project/Dependencies.scala]: I updated the rules re version upgrades. You can keep them in this PR :-)

File: web/app/TheApplicationLoader.scala (hunk 1 above)
Comment (omainegra) [inline web/app/TheApplicationLoader.scala]: Yes, it is. For instance

```scala
class TheApplicationLoader extends GuiceApplicationLoader {
  println("It's working!!!")
}
```

This effectively prints the message to console.

About why the it tests didn't. Let me try to figure it out

Comment (omainegra) [inline web/app/TheApplicationLoader.scala]: Well, actually this doesn't work:

```scala
class TheApplicationLoader extends GuiceApplicationLoader {
  println("It's working!!!")

  override protected def builder(context: ApplicationLoader.Context): GuiceApplicationBuilder = {
    ???
  }
}
```

Please see the stacktrace

```
! @74l5nhkp5 - Internal server error, for (GET) [/] ->
 
play.api.UnexpectedException: Unexpected exception[NotImplementedError: an implementation is missing]
        at play.core.server.DevServerStart$$anon$1.$anonfun$get$3(DevServerStart.scala:183)
        at play.core.server.DevServerStart$$anon$1$$Lambda$3487/767047103.apply(Unknown Source)
        at scala.Option.map(Option.scala:146)
        at play.core.server.DevServerStart$$anon$1.$anonfun$get$2(DevServerStart.scala:133)
        at play.core.server.DevServerStart$$anon$1$$Lambda$3486/280881128.apply(Unknown Source)
        at scala.util.Success.flatMap(Try.scala:247)
        at play.core.server.DevServerStart$$anon$1.$anonfun$get$1(DevServerStart.scala:131)
        at play.core.server.DevServerStart$$anon$1$$Lambda$3482/981851710.apply(Unknown Source)
        at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:653)
        at scala.concurrent.Future$$$Lambda$3483/977719985.apply(Unknown Source)
Caused by: scala.NotImplementedError: an implementation is missing
        at scala.Predef$.$qmark$qmark$qmark(Predef.scala:284)
        at TheApplicationLoader.builder(TheApplicationLoader.scala:8)
        at play.api.inject.guice.GuiceApplicationLoader.load(GuiceApplicationLoader.scala:21)
        at play.core.server.DevServerStart$$anon$1.$anonfun$get$6(DevServerStart.scala:171)
        at play.core.server.DevServerStart$$anon$1$$Lambda$3490/1749365538.apply(Unknown Source)
        at play.utils.Threads$.withContextClassLoader(Threads.scala:21)
        at play.core.server.DevServerStart$$anon$1.$anonfun$get$3(DevServerStart.scala:168)
        at play.core.server.DevServerStart$$anon$1$$Lambda$3487/767047103.apply(Unknown Source)
        at scala.Option.map(Option.scala:146)
        at play.core.server.DevServerStart$$anon$1.$anonfun$get$2(DevServerStart.scala:133)

```

Comment (ScalaWilliam) [inline web/app/TheApplicationLoader.scala]: Strange. Perhaps the integration test doesn't use the application loader actually. Hmm.

Comment (ScalaWilliam) [inline web/app/TheApplicationLoader.scala]: I don't wish to have two separate execution paths at all. There's no room for safe refactoring unless the integration test points out errors.

https://www.playframework.com/documentation/2.6.x/ScalaTestingWithComponents

Apologies, I did not foresee this at all.

Can you propose how to approach this?

**I'm signing off for the night now, 11pm**

Comment (omainegra) [inline web/app/TheApplicationLoader.scala]: I fixed the IT tests to use the new loader

Comment (ScalaWilliam) [inline web/app/TheApplicationLoader.scala]: Oh nice, I'll take a look in the morning!

File: web/it/TheApplicationFactory.scala [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import org.scalatestplus.play.FakeApplicationFactory
+import play.api.inject.DefaultApplicationLifecycle
+import play.api.{Application, ApplicationLoader, Configuration, Environment}
+import play.core.DefaultWebCommands
+
+/**
+  * Created by omainegra on 7/19/17.
+  */
+trait TheApplicationFactory extends FakeApplicationFactory {
+  override def fakeApplication: Application = {
```
Comment (ScalaWilliam) [inline web/it/TheApplicationFactory.scala]: A lot of this is done by `ApplicationLoader.createContext`.
Any reason for choosing this approach?

File: web/it/TheApplicationFactory.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import org.scalatestplus.play.FakeApplicationFactory
+import play.api.inject.DefaultApplicationLifecycle
+import play.api.{Application, ApplicationLoader, Configuration, Environment}
+import play.core.DefaultWebCommands
+
+/**
+  * Created by omainegra on 7/19/17.
```
Comment (ScalaWilliam) [inline web/it/TheApplicationFactory.scala]: Comment needed, such "For Compile-time DI".

File: web/it/TheApplicationFactory.scala (hunk 3 above)
Comment (omainegra) [inline web/it/TheApplicationFactory.scala]: None, just to have more control over test configuration. Changed

File: web/it/TheApplicationFactory.scala (hunk 4 above)
Comment (omainegra) [inline web/it/TheApplicationFactory.scala]: Done

**Pull request merged**